byterun/caml/hooks.h

```diff
@@ -17,22 +17,92 @@
 #define CAML_HOOKS_H
 
 #include "misc.h"
-#include "memory.h"
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 #ifdef CAML_INTERNALS
 
-#ifdef NATIVE_CODE
+/* Some macros to help declaring, exporting and using hooks */
+#define EXTERN_HOOK1(hook_name, arg1)      \
+  CAMLextern void (*hook_name)(arg1);
+#define DECLARE_HOOK1(hook_name, arg1)     \
+  CAMLexport void (*hook_name)(arg1) = NULL;
+#define MAYBE_HOOK1(hook_name, arg1) \
+  { if( hook_name != NULL ){ hook_name(arg1); } }
```

Comment:
Indeed -- this macro (and others in this PR) could be improved in several ways:
   * it should avoid evaluating `hook_name` twice
   * it [should be wrapped with `do .. while (0)`](http://c-faq.com/cpp/multistmt.html).
   * other macros (`EXTERN_HOOK1` etc.) should not end with a semicolon, since the caller also adds a terminating semicolon